ouroboros-consensus/docs/HardFork.md

```diff
@@ -0,0 +1,211 @@
+"rejected/discarded" status at some point.
+
+```
+active proposal
+    --> (sufficient votes)
+confirmed proposal
+    --> (2k slots later)
+stably-confirmed proposal
+    --> (sufficient endorsements)
+candidate proposal
+   --> (2k slots later)
+stably-candidated proposal    (Frisby: stably-nominated?)
+   --> (epoch transition)
+adopted proposal
+```
+
+### Initiating the hard fork
+
+Proposals to initiate the hard fork can be submitted and voted on before all
+core nodes are ready. After all, once a proposal is "stably-confirmed", it will
+effectively remain so indefinitely until nodes endorse it (or it gets superseded
+by another proposal). This means that nodes can vote to initiate the hard fork,
+_then_ wait for everybody to update their software, and once updated, the
+proposal is endorsed and eventually the hard fork is initiated.
+
+Endorsement is somewhat implicit. The node operator does not submit an explicit
+"endorsement transaction", but instead restarts the node (probably after a
+software update that makes the node ready to support the hard fork) with a new
+protocol version (as part of a config file or command line parameter), which
+then gets included in the blocks that the node produces (this value is part of
+the static `ByronConfig`: `byronProtocolVersion`).
+
+(Note that a node restart is necessary for _any_ change to a protocol parameter,
+even though most parameters do not require any change to the software at all.)
+
+### Software version
+
+The Byron header also records a software version (`headerSoftwareVersion`). This
+is a legacy concern only, and is present in but ignored by the current Byron
+implementation, and entirely absent from the Byron specification.
+
+## Shelley
+
+### Moment of the hard fork
+
+Similar to the Byron ledger, the Shelley ledger provides a "current protocol
+version", but it is a two-tuple (not a three-tuple), containing only a
+`hard fork` component and `soft fork` component:
+
+```haskell
+_protocolVersion :: (Natural, Natural)
+```
+
+in `PParams` (currently, module `PParams` in
+`chain-and-ledger/executable-spec/src/PParams.hs`).
+
+The hard fork from Shelley to its successor (Goguen?) will be initiated
+once the hard fork component of this version gets incremented.
+
+### The update mechanism for the protocol version
+
+The update mechanism in Shelley is simpler than it is in Byron. There is no
+distinction between votes and proposals: to "vote" for a proposal one merely
+submits the exact same proposal. There is also no separate endorsement step.
+The procedure is as follows:
+
+1. Proposals (change parameter `x` to value `y` are submitted to the chain)
+2. "Voting" (submitting of proposals) ends `2k` slots before the end of the
+   epoch (that `2k` might need to be adjusted pending more detailed analysis
+   from the researchers -- the spec records this as a single specification
+   parameter (not protocol parameter!) called `slotsPrior`).
```

Comment:
This is out of date, it's now `3k/f` instead of `2k` and `slotsPrior` is now called `stabilityWindow`, see https://github.com/input-output-hk/cardano-ledger-specs/pull/1424.

Reply:
It's actually _twice_ that :).